src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+            $response->setExpires($response->getDate());
+
+            if ($this->flagDirectives['no-store']) {
+                $response->headers->set('Cache-Control', 'no-cache, no-store, must-revalidate');
+            } else {
+                $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+            }
+
+            return;
+        }
+
+        $flags = array_filter($this->flagDirectives);
+
+        if (isset($flags['must-revalidate'])) {
+            $flags['no-cache'] = true;
         }
 
-        if (!$response->isFresh() || !$response->isCacheable()) {
-            $this->cacheable = false;
+        $response->headers->set('Cache-Control', implode(', ', array_keys($flags)));
+
+        $maxAge = null;
+        $sMaxage = null;
+
+        if ($this->ageDirectives['max-age']) {
+            $maxAge = $this->ageDirectives['max-age'] + $this->age;
+            $response->headers->addCacheControlDirective('max-age', $maxAge);
         }
 
-        if (!$this->cacheable) {
-            $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+        if ($this->ageDirectives['s-maxage']) {
+            $sMaxage = $this->ageDirectives['s-maxage'] + $this->age;
 
-            return;
+            if ($maxAge !== $sMaxage) {
+                $response->headers->addCacheControlDirective('s-maxage', $sMaxage);
+            }
+        }
+
+        if ($this->ageDirectives['expires']) {
+            $response->setExpires($response->getDate()->modify('+'.$this->ageDirectives['expires'].' seconds'));
         }
+    }
 
-        $this->ttls[] = $response->getTtl();
-        $this->maxAges[] = $response->getMaxAge();
+    /**
+     * RFC2616, Section 13.4.
+     *
+     * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.4
+     *
+     * @return bool
+     */
+    private function willMakeFinalResponseUncacheable(Response $response)
+    {
+        // RFC2616: A response received with a status code of 200, 203, 300, 301 or 410
+        // MAY be stored by a cache […] unless a cache-control directive prohibits caching.
+        if ($response->headers->hasCacheControlDirective('no-cache')
+            || $response->headers->getCacheControlDirective('no-store')
+        ) {
+            return true;
+        }
 
-        if ($this->isNotCacheableResponseEmbedded) {
-            $response->headers->removeCacheControlDirective('s-maxage');
-        } elseif (null !== $maxAge = min($this->maxAges)) {
-            $response->setSharedMaxAge($maxAge);
-            $response->headers->set('Age', $maxAge - min($this->ttls));
+        // Last-Modified and Etag headers cannot be merged, they render the response uncacheable
+        // by default (except if the response also has max-age etc.).
+        if (\in_array($response->getStatusCode(), array(200, 203, 300, 301, 410))
+            && null === $response->getLastModified()
+            && null === $response->getEtag()
```

Comment:
What is missing here is that *having* a Etag or LastModified on an embedded response must make the final response uncacheable if it does not have a max-age or expire (as the validation cannot be done).

Reply:
I agree it's a bit confusing, but thats is what the code does.

The method returns `true` if the final response is uncacheable. So
 1. If the status code is cacheable by default, **and** we do not have a `Last-Modified` or `E-Tag`, return `false` because the response is cacheable
 2. If the status code was invalid or one of the headers was present, only return `false` (= cacheable) if one of the cache headers (`max-age` etc.) is present.
 3. Finally return `true` (= uncacheable) if none of the conditions apply.

Reply:
even if there are max-age/expire, the etag / last-modified can still not be validated. 

or does http cache on if-none-match/if-modified-since requests check all embedded ESI too? i think not, as in the main method, we remove etag and last-modified from the master response we send. i think this can be simplified to "if there is any etag / last-modified then the master response is not cacheable"

Reply:
The cache cannot validate ESI responses. But removing ETag or Last-Modified only prevents the response from being validated, it can still be cached for the `max-age` etc. duration. Validation only kicks in **after** `max-age` etc. has expired.

Reply:
true that, agree. see my suggestion above for how to restructure the logic to make it easier to understand.